File: src/agent/agent.go
package agent

import (
	"context"
	"discord-gemini-bot/src/models"
	"discord-gemini-bot/src/prompts"
	"discord-gemini-bot/src/tools"
	"discord-gemini-bot/src/types"
	"fmt"
	"log"
	"regexp"
	"strings"
)

// Agent represents the main agent class that handles user interactions
type Agent struct {
	model  models.LLMModel
	memory *types.ConversationMemory
	tools  map[string]tools.Tool
}

// NewAgent creates a new agent instance
func NewAgent(model models.LLMModel, memory *types.ConversationMemory, toolList []tools.Tool) *Agent {
	// Create tools map
	toolsMap := make(map[string]tools.Tool)
	for _, tool := range toolList {
		toolsMap[tool.Name()] = tool
	}

	agent := &Agent{
		model:  model,
		memory: memory,
		tools:  toolsMap,
	}

	// Set up system prompt
	toolsString := agent.getToolsString()
	toolNames := agent.getToolNames()
	systemPrompt := fmt.Sprintf(prompts.GetAgentSystemPromptTemplate(), toolsString, toolNames)
	model.SetSystemPrompt(systemPrompt)

	log.Printf("System prompt set for agent")
	return agent
}

// getToolNames returns a comma-separated string of tool names
func (a *Agent) getToolNames() string {
	var names []string
	for name := range a.tools {
		names = append(names, name)
	}
	return strings.Join(names, ", ")
}

// getToolsString returns a formatted string of all available tools
func (a *Agent) getToolsString() string {
	var toolDescriptions []string
	for name, tool := range a.tools {
		toolDescriptions = append(toolDescriptions, fmt.Sprintf("%s: %s", name, tool.Description()))
	}
	return strings.Join(toolDescriptions, "\n")
}

// GetResponse gets a response from the agent
func (a *Agent) GetResponse(ctx context.Context, author, message string, images []map[string]interface{}) (string, error) {
	// Add user message to memory
	a.memory.AddMessage(author, message)
	
	// Get conversation history
	messages := a.memory.GetHistory()
	
	// Generate response using the model
	response, err := a.model.GenerateWithHistoryAsync(ctx, messages)
	if err != nil {
		return "", fmt.Errorf("error generating response: %w", err)
	}

	log.Printf("Model's raw response: %s", response)

	// Check for tool use
	toolRegex := regexp.MustCompile(`Action: (\w+)\nAction Input: (.*)`)
	matches := toolRegex.FindStringSubmatch(response)
	
	if len(matches) >= 3 {
		toolName := strings.TrimSpace(matches[1])
		toolInput := strings.TrimSpace(matches[2])
		
		log.Printf("Tool use detected: %s with input %s", toolName, toolInput)
		
		tool, exists := a.tools[toolName]
		if exists {
			// Execute the tool
			toolResult, err := tool.ARun(ctx, toolInput)
			if err != nil {
				log.Printf("Error executing tool %s: %v", toolName, err)
				observation := fmt.Sprintf("Tool %s failed: %v", toolName, err)
				a.memory.AddMessage("AI", observation)
			} else {
				observation := fmt.Sprintf("Tool %s used. Observation: %s", toolName, toolResult.ReturnDisplay)
				log.Printf("Tool observation: %s", observation)
				a.memory.AddMessage("AI", observation)
				
				// Get a new response with the tool's output
				history := a.memory.GetHistory()
				prompt := fmt.Sprintf("%s\n\nTOOLS:\n------\n%s\n\nPrevious conversation history:\n%s\n\nNew input: %s: %s\nFinal Answer:", 
					prompts.GetAgentSystemPromptTemplate(), 
					a.getToolsString(), 
					a.formatHistory(history), 
					author, 
					message)
				
				log.Printf("Prompt sent to model after tool use")
				
				// Generate follow-up response
				response, err = a.model.GenerateAsync(ctx, prompt, images)
				if err != nil {
					return "", fmt.Errorf("error generating follow-up response: %w", err)
				}
				
				log.Printf("Model's raw response after tool use: %s", response)
			}
		} else {
			log.Printf("Tool %s not found", toolName)
		}
	}

	// Add AI response to memory
	a.memory.AddMessage("AI", response)
	
	log.Printf("Agent's final response: %s", response)
	return response, nil
}

// formatHistory formats the conversation history for the prompt
func (a *Agent) formatHistory(messages []*types.Message) string {
	var formatted []string
	for _, msg := range messages {
		formatted = append(formatted, fmt.Sprintf("%s: %s", msg.Role, msg.Content))
	}
	return strings.Join(formatted, "\n")
}

File: python_deprecated/src/agent.py
import re
import logging
from types.memory import ConversationMemory
from prompts import get_agent_system_prompt_template
from tools.tools import Tool
from models.llm_model import LLMModel
from typing import List, Dict, Any

logger = logging.getLogger(__name__)

class Agent:
    """The main agent class that handles user interactions."""

    def __init__(self, model: LLMModel, memory: ConversationMemory, tools: List[Tool]):
        self.model = model
        self.memory = memory
        self._tools = {tool.name: tool for tool in tools}

        tools_string = self._get_tools_string()
        tool_names = self._get_tool_names()
        system_prompt = get_agent_system_prompt_template().format(tools=tools_string, tool_names=tool_names)
        self.model.set_system_prompt(system_prompt)
        logger.debug(f"System prompt set: {system_prompt}")

    def _get_tool_names(self) -> str:
        """Returns a comma-separated string of tool names."""
        return ', '.join(self._tools.keys())

    def _get_tools_string(self) -> str:
        """Returns a formatted string of all available tools."""
        return "\n".join([
            f'{tool.name}: {tool.description}' for tool in self._tools.values()
        ])

    async def get_response(self, author: str, message: str, images: List[Dict[str, Any]] = None) -> str:
        """Gets a response from the agent."""
        self.memory.add_message(author, message)
        messages = self.memory.get_history()
        
        # Use LLMModel interface
        response = await self.model.generate_with_history_async(messages)
        
        logger.info(f"Model's raw response: {response}")

        # Check for tool use
        tool_match = re.search(r"Action: (\w+)\nAction Input: (.*)", response)
        if tool_match:
            tool_name = tool_match.group(1)
            tool_input = tool_match.group(2)
            logger.info(f"Tool use detected: {tool_name} with input {tool_input}")
            tool = self._tools.get(tool_name)
            if tool:
                tool_result = await tool.arun(tool_input)
                observation = f"Tool {tool_name} used. Observation: {tool_result.return_display}"
                logger.info(f"Tool observation: {observation}")
                self.memory.add_message("AI", observation)
                
                # Get a new response with the tool's output
                history = self.memory.get_history()
                prompt = f"{get_agent_system_prompt_template()}\n\nTOOLS:\n------\n{tools_string}\n\nPrevious conversation history:\n{history}\n\nNew input: {author}: {message}\nFinal Answer:"
                logger.debug(f"Prompt sent to model after tool use: {prompt}")
                
                # Use LLMModel interface for follow-up response
                response = await self.model.generate_async(prompt, images=images)
                
                logger.info(f"Model's raw response after tool use: {response}")

        self.memory.add_message("AI", response)
        logger.info(f"Agent's final response: {response}")
        return response